[Screen Time] Remove Website Data only removes data for top-level domains, not absolute URLs

jesxilin opened the pull request:
#### e2667f26ff6238268702229fb1dfad7c548f10b1
<pre>
[Screen Time] Remove Website Data only removes data for top-level domains, not absolute URLs
<a href="https://bugs.webkit.org/show_bug.cgi?id=294337">https://bugs.webkit.org/show_bug.cgi?id=294337</a>
<a href="https://rdar.apple.com/153113708">rdar://153113708</a>

Reviewed by Aditya Keerthi.

Currently, when removing website data, we only supply top-level domains
to be removed. This does not remove absolute URLs, thus not fully
removing all the URLs associated with a particular domain.

Implement the ScreenTime.RemoveData API test and remove the
FIXME comment.

* Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
(WebKit::ScreenTimeWebsiteDataSupport::removeScreenTimeData):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(TEST(ScreenTime, RemoveData)):

Canonical link: <a href="https://commits.webkit.org/296209@main">https://commits.webkit.org/296209@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a488be1b1a1feb293c1e3cdc75826a296cbade5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107713 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27387 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17802 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112923 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109668 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28080 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35916 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81782 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110647 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22257 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97074 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62170 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/21690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15217 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57685 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91628 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15251 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116044 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34789 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25635 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90820 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35166 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93326 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90579 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35493 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13257 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/30544 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17417 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34694 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/40249 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34439 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37801 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36102 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -88,8 +88,25 @@ void removeScreenTimeData(const HashSet<URL>& websitesToRemove, const WebsiteDat
-    for (auto& url : websitesToRemove)
-        [webHistory deleteHistoryForURL:url.createNSURL().get()];
+    // Convert websitesToRemove to NSStrings.
+    NSMutableSet<NSString*> *websitesToRemoveDomains = [NSMutableSet set];
```
```suggestion
    RetainPtr websitesToRemoveDomains = [NSMutableSet set];
```

lilyspiniolas commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -88,8 +88,25 @@ void removeScreenTimeData(const HashSet<URL>& websitesToRemove, const WebsiteDat
+    NSMutableSet<NSString*> *websitesToRemoveDomains = [NSMutableSet set];
+    for (auto& url : websitesToRemove) {
+        RetainPtr nsurl = url.createNSURL();
+        [websitesToRemoveDomains addObject:[nsurl absoluteString]];
```
can this all just be a single line?

lilyspiniolas replied:
*the inner portion of the loop

jesxilin replied:
Should be!

lilyspiniolas commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -88,8 +88,25 @@ void removeScreenTimeData(const HashSet<URL>& websitesToRemove, const WebsiteDat
+        if (error)
+            return;
+
+        for (NSURL * site : urls) {
```
```suggestion
        for (NSURL *site : urls) {
```

lilyspiniolas commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -88,8 +88,25 @@ void removeScreenTimeData(const HashSet<URL>& websitesToRemove, const WebsiteDat
+        for (NSURL * site : urls) {
+            for (NSString* domainString : websitesToRemoveDomains) {
+                if ([[site absoluteString] containsString:domainString])
+                    [webHistory deleteHistoryForURL:[NSURL URLWithString: [site absoluteString]]];
```
```suggestion
                    [webHistory deleteHistoryForURL:[NSURL URLWithString:[site absoluteString]]];
```

lilyspiniolas commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -88,8 +88,25 @@ void removeScreenTimeData(const HashSet<URL>& websitesToRemove, const WebsiteDat
+            return;
+
+        for (NSURL * site : urls) {
+            for (NSString* domainString : websitesToRemoveDomains) {
```
```suggestion
            for (NSString *domainString : websitesToRemoveDomains) {
```

rr-codes commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -88,8 +88,25 @@ void removeScreenTimeData(const HashSet<URL>& websitesToRemove, const WebsiteDat
+    // Convert websitesToRemove to NSStrings.
+    NSMutableSet<NSString*> *websitesToRemoveDomains = [NSMutableSet set];
+    for (auto& url : websitesToRemove) {
+        RetainPtr nsurl = url.createNSURL();
```
this could possibly return nil, and therefore cause a crash on the line below

pxlcoder replied:
```suggestion
        RetainPtr nsURL = url.createNSURL();
```

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -88,8 +88,25 @@ void removeScreenTimeData(const HashSet<URL>& websitesToRemove, const WebsiteDat
+
+        for (NSURL * site : urls) {
+            for (NSString* domainString : websitesToRemoveDomains) {
+                if ([[site absoluteString] containsString:domainString])
```
We should compare to the `host`.

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -88,8 +88,22 @@ void removeScreenTimeData(const HashSet<URL>& websitesToRemove, const WebsiteDat
+        for (NSURL *url : urls) {
+            for (NSString *domainString : websitesToRemoveDomains.get()) {
+                if ([[url host] hasSuffix:domainString])
+                    [webHistory deleteHistoryForURL:[NSURL URLWithString:[url absoluteString]]];
```
```suggestion
                    [webHistory deleteHistoryForURL:url];
```

`[NSURL URLWithString:[url absoluteString]]` just gets you the same URL.

pxlcoder commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
```diff
@@ -88,8 +88,22 @@ void removeScreenTimeData(const HashSet<URL>& websitesToRemove, const WebsiteDat
+            return;
+
+        for (NSURL *url : urls) {
+            for (NSString *domainString : websitesToRemoveDomains.get()) {
```
```suggestion
        for (NSURL *url in urls) {
            for (NSString *domainString in websitesToRemoveDomains.get()) {
```

(since Objective-C)

jesxilin committed:
[Screen Time] Remove Website Data only removes data for top-level domains, not absolute URLs
https://bugs.webkit.org/show_bug.cgi?id=294337
rdar://153113708

Reviewed by Aditya Keerthi.

Currently, when removing website data, we only supply top-level domains
to be removed. This does not remove absolute URLs, thus not fully
removing all the URLs associated with a particular domain.

Implement the ScreenTime.RemoveData API test and remove the
FIXME comment.

* Source/WebKit/UIProcess/WebsiteData/Cocoa/ScreenTimeWebsiteDataSupport.mm:
(WebKit::ScreenTimeWebsiteDataSupport::removeScreenTimeData):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScreenTime.mm:
(TEST(ScreenTime, RemoveData)):

Canonical link: https://commits.webkit.org/296209@main
